seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,185 @@
+            r = request_func(url, headers = headers, **kwargs)
+        if r.status_code != 200:
+            logger.info('{} {} {} {}'.format(method.upper(), url, r.status_code, len(r.text)))
+            raise TerraAPIException('Error: called Terra API "{}" got status: {} with a reason: {}'.format(
+                path, r.status_code, r.reason))
+        return json.loads(r.text)
+
+    def get(self, path, headers=None, root_url=None, **kwargs):
+        """
+        Call Terra API with HTTP GET method with an authentication header.
+
+        :param
+            path: A string of API path (start right after the domain name without leading slash (/)
+            headers (dict): Include additional headers as key-value pairs
+            root_url: the url up to the domain name ending with a slash (/)
+            kwargs: other parameters for the HTTP call, e.g. queries.
+        :return
+            HTTP response
+        """
+        return self._make_request('get', path, headers, root_url, **kwargs)
+
+
+_service_account_session = ServiceAccountSession()
+
+
+def sa_get_workspace_acl(workspace_namespace, workspace_name):
+    """
+    Request FireCloud access control list for workspace with a service account (sa).
+
+    Args:
+        workspace_namespace (str): namespace (name of billing project) of the workspace
+        workspace_name (str): the name of the workspace
+    Returns:
+        {
+            "user1Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user2Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user3Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            }
+          }
+          :param workspace_name:
+          :param workspace_namespace:
+    """
+    uri = "api/workspaces/{0}/{1}/acl".format(workspace_namespace, workspace_name)
+    r = _service_account_session.get(uri)
+    return r.get('acl', {})
+
+
+def _get_social_access_token(user):
+    social = user.social_auth.get(provider = 'google-oauth2')
+    if (social.extra_data['auth_time'] + social.extra_data['expires'] - 10) <= int(
+            time.time()):  # token expired or expiring?
+        strategy = load_strategy()
+        try:
+            social.refresh_token(strategy)
+        except Exception as ee:
+            logger.info('Refresh token failed. {}'.format(str(ee)))
+            raise TerraAPIException('Refresh token failed. {}'.format(str(ee)))
+    return social.extra_data['access_token']
+
+
+def anvil_call(method, path, user=None, headers=None, root_url=None, access_token=None, **kwargs):
+    if user:
+        access_token = _get_social_access_token(user)
+
```

Comment:
so this is a function that can be called in 2 pretty fundamentally different ways - with a user where you need to get the access token, or with an access token already provided. But because everything is optional arguments and there isn't good argument validation, its really easy to mess up using this function, either by providing both user and token or buy providing neither, and in both cases you would get weird undefined behavior. I think it would be better to make `anvil_call` require `access_token` as a positional argument, and then write another function for anvil calls with a user that calls it
```
def user_anvil_call(user, method, path, **kwargs):
   access_token = _get_social_access_token(user)
   return anvil_call(method, path, access_token, **kwargs)
```

Reply:
Good point.